paddle/go/optimizer/parameter_optimizer.cc

```diff
@@ -0,0 +1,67 @@
+#include "parameter_optimizer.h"
+#include "optimizer_factory.h"
+#include <glog/logging.h>
+
+namespace paddle {
+namespace optimizer {
+
+template <class T>
+ParameterOptimizer<T>* ParameterOptimizer<T>::create(
+    const ::std::string &config_proto) {
+  paddle::OptimizerConfig config;
+  CHECK(config.ParseFromString(config_proto) == 0) << "error : optimizer config";
+  CHECK(config_valid(config) == 0) << "error : invalid optimizer config ";
+  ParameterOptimizer<T> *opt = nullptr;
+  switch (config.optimizer_name()) {
+  case "SGD" : opt = new SGDOptimizer<T>(config); break;
+  case "Adagrad" : opt = new AdagradOptimizer<T>(config); break;
+  case "Adadelta" : opt = new AdadeltaOptimizer<T>(config); break;
+  case "Adam" : opt = new AdamOptimizer<T>(config); break;
+  default:
+    opt = new SGDOptimizer<T>(config);
+  }
+  return opt;
+}
+
+template<class T>
+double ParameterOptimzier<T>::get_learning_rate() {
+  if (config_.lr_type() == paddle::OptimizerConfig_LearningRateType_Linear) {
+    learning_rate = ::std::max(learning_rate - lr_decay_a * num_sample_passed, lr_decay_b);
```

Comment:
I think the plain simple SGD optimizer only have constant learning rate and does not contain a learning rate schedule. This learning rate schedule function should not be in the base class.

I think in v1 we don't need learning rate schedule. And in future, we maybe need to use composition instead of inheritance for learning rate schedule. Reference: https://stackoverflow.com/a/49016/852385

> Think of containment as a has a relationship. A car "has an" engine, a person "has a" name, etc.
Think of inheritance as an is a relationship. A car "is a" vehicle, a person "is a" mammal, etc.